authentication_executor/authenticators/oauth2_authenticator.py

```diff
@@ -0,0 +1,157 @@
+            if redirect_location is None:
+                self.debug_data += ['No `location` header in authorization code response']
+                raise BadAuthorizationCodeResponse()
+            parsed_location = urlparse(redirect_location)
+            parsed_location_query = parse_qs(parsed_location.query)
+
+            authorization_code = parsed_location_query.get('code')
+            if authorization_code is None:
+                self.debug_data += ['No authorization code found in authorization code response']
+                raise BadAuthorizationCodeResponse()
+            return authorization_code
+
+        except Exception as e:
+            self.debug_data += ["Authorization code request failed", str(e)]
+            raise e
+
+    def call_token_url(self, requests_session, authorization_code):
+        try:
+            resp = requests_session.request(
+                method='POST',
+                url=self.execution_spec.token_url,
+                data=self.get_auth_request_body(authorization_code),
+                json=None,
+                headers=self.get_auth_request_headers()
+            )
+            resp.raise_for_status()
+            token_response = resp.json()
+            if token_response.get('token_type', 'bearer').lower() != 'bearer':
+                self.debug_data += [f'Unsupported token type {token_response.get("token_type")}']
+                raise UnsupportedTokenTypeException()
+
+            return self.extract_token_from_oauth_response(token_response)
+        except Exception as e:
+            self.debug_data += ["Token request failed", str(e)]
+            raise e
+
+    def get_auth_request_body(self, authorization_code=None):
+        return get_without_none_values({
+            'grant_type': self.execution_spec.grant_type,
+            'client_id': self.execution_spec.client_id,
+            'client_secret': self.execution_spec.client_secret,
+            'username': self.execution_spec.username,
+            'password': self.execution_spec.password,
+            'scope': self.execution_spec.scope,
+            'redirect_uri': self.execution_spec.callback_url,
+            'state': self.execution_spec.state,
+            'code': authorization_code
+        })
+
+    def get_auth_request_headers(self):
+        request_headers = {
+            'Content-Type': 'application/x-www-form-urlencoded',
+            'Accept': 'application/json'
+        }
+        if self.execution_spec.client_creds_in_headers:
+            request_headers['Authorization'] = 'Basic ' + base64.b64encode(f'{self.execution_spec.client_id}:{self.execution_spec.client_secret}'.encode()).decode('utf-8')
+        return request_headers
+
+    @staticmethod
+    def extract_token_from_oauth_response(response_json):
+        token = response_json.get('access_token', response_json.get('accessToken', response_json.get('token')))  # not everyone follows spec perfectly sadly
```

Comment:
maybe there is a way to search for lower case "token" in keys and take it? 
still not the best option but maybe it will be nicer codewise

Reply:
This response contains a 'token_type' field and may contain 'refresh_token' as well as other unexpected fields, so im not sure if a fuzzy search is safe here

Reply:
ok. I I understand